GitExtensions/StaticDependencyInjection.cs

```diff
@@ -0,0 +1,118 @@
+﻿using GitCommands;
+using System;
+using System.IO;
+using System.Collections.Concurrent;
+
+namespace GitExtensions
+{
+
+    public class DIObjectInstance<T> where T : class, new()
+    {
+        private static object lockObj = new object();
+        private T _Inst;
+        public T Inst
+        {
+            get
+            {
+                if (_Inst == null)
+                {
+                    lock (lockObj)
+                    {
+                        if (_Inst == null)
+                        {
+                            _Inst = StaticDI.CreateInstance<T>(ClearInst);
+                        }
+                    }
+                }
+
+                return _Inst;
+            }
+        }
+
+        private void ClearInst()
+        {
+            lock (lockObj)
+            {
+                _Inst = null;
+            }
+        }
+    }
+
+    public interface IInstanceFactory
+    {
+        T CreateInstance<T>() where T : class, new();
+    }
+
+    public class NewInstanceFactory : IInstanceFactory
+    {
+        public T CreateInstance<T>() where T : class, new()
+        {
+            return new T();
+        }
+    }
+
+    public class StaticDI
```

Comment:
Apologies, the following post turned out to be a long read... 📚 😪 
Hopefully it won't bore you to death 😄 


To me a global variable is essentially a side effect - changes outside my class may affect its behavior.
It is also much harder to reason about code dependencies and coupling and refactor the code. 

The "external" design also carries a significant overhead - instead of injecting an instance of `IFileSystem` interface and use its methods I need to "implement" all methods I need to call.
The API, as I said earlier, are also very un-aesthetic.


If I understand your comment correctly, you don't have much experience with IoC containers (please forgive me if I misinterpreted it).
If we were to use a IoC container the structure of the code would be something like the following. 

Let's set the context:

* There are DTOs (or POCOs if you like):

```
    internal class CommandLineArgs
    {
        public CommandLineArgs(string command, string workingDir)
        {
            Command = command ?? string.Empty;
            WorkingDir = workingDir ?? string.Empty;
        }

        public string Command { get; }
        public string WorkingDir { get; }
    }
```

DTOs (generally with few exceptions) define properties only and carry no behavioural implementations. Generally you will make them immutable too (exception for serializable DTOs).
Because DTOs have no behaviours they are generally not injected in constructors but passed as input/output parameters into methods.

* There are "worker" classes - these exhibit behaviours all of which should be linked to a single responsibility (i.e. file reader/writer; settings manager - load, get, save settings; git remotes controller etc).
You make each of these classes interface-based and each class has all its dependencies injected via its constructor:

```
    internal interface ICommandLineArgsParser
    {
        CommandLineArgs Parse(string[] args);
    }

    internal sealed class CommandLineArgsParser : ICommandLineArgsParser
    {
        private readonly IWorkingPathProvider _workingPathProvider;

        public CommandLineArgsParser(IWorkingPathProvider workingPathProvider)
        {
            _workingPathProvider = workingPathProvider;
        }

        public CommandLineArgs Parse(string[] args) { ... }
    }
```

* There is a "registrar" class in which we define type registrations.
Here I use Autofac as an example (because I am more familiar with it), but the approach works for any other IoC container, different syntax. 

```
var builder = new ContainerBuilder();
 
builder.RegisterType<FileSystem>().As<IFileSystem>().As;
builder.RegisterType<WorkingPathProvider>().As<IWorkingPathProvider>();
builder.RegisterType<CommandLineArgsParser>().As<ICommandLineArgsParser>();
// any other registrations
 
var container = builder.Build();
```

You can also define the lifetime scoping of each instance as well, generally stateless classes can be made singletons - this way you improve the performance of the app as well.

This registrar class will be executed at the start of the program, it will build the dependency tree. For classes which we switch to DI pattern the IoC container will inject the required type instances. For classes which aren't converted to DI-pattern we will use the service locator-like approach (SL is an anti-pattern, but we can use it as a mean to get to the target state. Also [read this](http://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/)):

```
// Program.cs static Main() - this is more an edge case scenario

            string[] args = Environment.GetCommandLineArgs();
            var parser = ServiceLocator.Resolve<ICommandLineArgsParser>(); // <-- using it as anti-pattern
            var commandLineArgs = parser.Parse(args);

            if (!string.IsNullOrWhiteSpace(commandLineArgs.WorkingDir))
            {
                uiCommands = new GitUICommands(commandLineArgs.WorkingDir);
            }


// a more typical example:
    public sealed partial class FormCommit : GitModuleForm
    {
        private readonly ICommitTemplateManager _commitTemplateManager;

        public FormCommit(GitUICommands aCommands, CommitKind commitKind, GitRevision editedCommit)
            : base(true, aCommands)
        {
           // arguably here we aren't using it as a SL anti-pattern, more like DI
           // this must be treated as a temporary bandaid and
           // will have to be changed to proper DI at some point
            _commitTemplateManager = ServiceLocator.Resolve<ICommitTemplateManager>();
        }

        ...
    }
```